windows/src/desktop/insthelp/main.pas

```diff
@@ -46,7 +46,22 @@ implementation
   Winapi.Windows,
   Keyman.System.InstHelp.KeymanStartTaskUninstall,
   klog,
-  RegistryKeys;
+  RegistryKeys,
+  UserMessages;
+
+procedure ShutdownKeyman;
+var
+  hwnd: THandle;
+const
+  KMC_StopProduct = 1;  // from UfrmKeyman7Main.pas
+begin
+  hwnd := FindWindow('TfrmKeyman7Main', nil);
+  if hwnd <> 0 then
+  begin
+    PostMessage(hwnd, WM_USER_ParameterPass, KMC_StopProduct, 0);
+    Sleep(1000); // We'll wait 1 second to give Keyman Engine time to shut down
```

Comment:
Why not a polling loop?

I don't know Delphi, so something like the following pseudocode:

```
var
  iterationsUntilFail = 10;

  // ...

while hwnd <> 0 && iterationsUntilFail > 0
  begin
    PostMessage(hwnd, WM_USER_ParameterPass, KMC_StopProduct, 0);
    Sleep(1000);
    iterationsUntilFail--;
    hwnd := FindWindow('TfrmKeyman7Main', nil);
end;

// loop auto-exits after enough iterations have passed.

Reply:
Right now, it looks like you're assuming that Keyman will always exit within that 1000ms timeframe without issue... which is a bit of an assumption.

Reply:
Yep. So is 10 seconds, of course, but it's a bit more graceful. Internally, the destruction of the window does not guarantee that the process has shut down, but in normal circumstances, it will be gone very shortly thereafter. I think this is fine. In abnormal circumstances, well, not much we can do but let them restart...